src/graphene.cpp

```diff
@@ -0,0 +1,1471 @@
+                    nSizeBlock, pfrom->GetLogName());
+
+                graphenedata.UpdateFilter(grapheneBlock.pGrapheneSet->GetFilterSerializationSize());
+                graphenedata.UpdateIblt(grapheneBlock.pGrapheneSet->GetIbltSerializationSize());
+                graphenedata.UpdateRank(grapheneBlock.pGrapheneSet->GetRankSerializationSize());
+                graphenedata.UpdateGrapheneBlock(nSizeGrapheneBlock);
+                graphenedata.UpdateAdditionalTx(grapheneBlock.GetAdditionalTxSerializationSize());
+            }
+        }
+        catch (const std::runtime_error &e)
+        {
+            pfrom->PushMessage(NetMsgType::BLOCK, *pblock);
+            LOG(GRAPHENE,
+                "Sent regular block instead - encountered error when creating graphene block for peer %s: %s\n",
+                pfrom->GetLogName(), e.what());
+        }
+    }
+    else
+    {
+        dosMan.Misbehaving(pfrom, 100);
+
+        return;
+    }
+
+    pfrom->blocksSent += 1;
+}
+
+bool IsGrapheneBlockValid(CNode *pfrom, const CBlockHeader &header)
+{
+    // check block header
+    CValidationState state;
+    if (!CheckBlockHeader(header, state, true))
+    {
+        return error("Received invalid header for graphene block %s from peer %s", header.GetHash().ToString(),
+            pfrom->GetLogName());
+    }
+    if (state.Invalid())
+    {
+        return error("Received invalid header for graphene block %s from peer %s", header.GetHash().ToString(),
+            pfrom->GetLogName());
+    }
+
+    return true;
+}
+
+bool HandleGrapheneBlockRequest(CDataStream &vRecv, CNode *pfrom, const CChainParams &chainparams)
+{
+    if (!pfrom->GrapheneCapable())
+    {
+        dosMan.Misbehaving(pfrom, 100);
+        return error("Graphene block message received from a non graphene block node, peer=%d", pfrom->GetId());
+    }
+
+    // Check for Misbehaving and DOS
+    // If they make more than 20 requests in 10 minutes then disconnect them
+    {
+        LOCK(cs_vNodes);
+        if (pfrom->nGetGrapheneLastTime <= 0)
+            pfrom->nGetGrapheneLastTime = GetTime();
+        uint64_t nNow = GetTime();
+        pfrom->nGetGrapheneCount *= std::pow(1.0 - 1.0 / 600.0, (double)(nNow - pfrom->nGetGrapheneLastTime));
+        pfrom->nGetGrapheneLastTime = nNow;
+        pfrom->nGetGrapheneCount += 1;
+        LOG(GRAPHENE, "nGetGrapheneCount is %f\n", pfrom->nGetGrapheneCount);
+        if (chainparams.NetworkIDString() == "main") // other networks have variable mining rates
+        {
+            if (pfrom->nGetGrapheneCount >= 20)
+            {
+                dosMan.Misbehaving(pfrom, 100); // If they exceed the limit then disconnect them
+                return error("sending too many GET_GRAPHENE messages");
+            }
+        }
+    }
+
+    CMemPoolInfo mempoolinfo;
+    CInv inv;
+    vRecv >> inv >> mempoolinfo;
+    graphenedata.UpdateInBoundMemPoolInfo(::GetSerializeSize(mempoolinfo, SER_NETWORK, PROTOCOL_VERSION));
+
```

Comment:
If mempoolinfo is a huge or number what happens?  I traced it to throwing an exception in CGrapheneSet.OptimalSymDiff.  Since we catch that exception it should be ok, but to reduce the attack analysis surface (and protect subsequent changes) let's block values greater than LARGE_MEM_POOL_SIZE right here.

If mempoolinfo is 0 what happens?

Reply:
I'll limit the upper range for mempool size during deserialization as you suggested below. Regarding the lower limit, I can confirm from testing on mainnet that passing mempool size = 0 does not produce unexpected behavior. Because it is an underestimate, graphene simply returns a Bloom filter with very high FPR and the IBLT fails to decode. So the end result is that an XThin block is sent after the graphene failure.